cpp/src/arrow/scalar_test.cc

```diff
@@ -930,117 +930,134 @@ TEST(TestDictionaryScalar, Cast) {
   }
 }
 
-TEST(TestSparseUnionScalar, Basics) {
-  auto ty = sparse_union({field("string", utf8()), field("number", uint64())});
+void CheckGetValidUnionScalar(const Array& arr, int64_t index, const Scalar& expected,
+                              const Scalar& expected_value) {
+  ASSERT_OK_AND_ASSIGN(auto scalar, arr.GetScalar(index));
+  ASSERT_TRUE(scalar->Equals(expected));
+
+  const auto& as_union = checked_cast<const UnionScalar&>(*scalar);
+  ASSERT_TRUE(as_union.is_valid);
+  ASSERT_TRUE(as_union.value->Equals(expected_value));
+}
 
-  auto alpha = MakeScalar("alpha");
-  auto beta = MakeScalar("beta");
-  ASSERT_OK_AND_ASSIGN(auto two, MakeScalar(uint64(), 2));
+void CheckGetNullUnionScalar(const Array& arr, int64_t index) {
+  ASSERT_OK_AND_ASSIGN(auto scalar, arr.GetScalar(index));
+  ASSERT_TRUE(scalar->Equals(MakeNullScalar(arr.type())));
 
-  auto scalar_alpha = SparseUnionScalar(alpha, ty);
-  auto scalar_beta = SparseUnionScalar(beta, ty);
-  auto scalar_two = SparseUnionScalar(two, ty);
+  const auto& as_union = checked_cast<const UnionScalar&>(*scalar);
+  ASSERT_FALSE(as_union.is_valid);
+  // XXX in reality, the union array doesn't have a validity bitmap.
+  // Validity is inferred from the underlying child value, which should maybe
+  // be reflected here...
+  ASSERT_EQ(as_union.value, nullptr);
```

Comment:
In this case, I think allowing UnionScalar::value to be nullptr instead of MakeNullScalar(...) is tantamount to considering shared_ptr<Scalar>(nullptr) to be an acceptable way to represent a null scalar. I think we should instead maintain scalars which are "shaped" as identically to their corresponding arrays as possible, and in this case I think that requires a non-nullptr UnionScalar::value.
```suggestion
  ASSERT_TRUE(as_union.value->is_valid);
```

Reply:
If we want scalars to be "shaped" closely to their corresponding arrays, do you think `as_union.is_valid` should be true here?
(note that `as_union.valid->is_valid` would be false, unlike your suggestion)

Note however that this PR doesn't change the current behaviour.

Reply:
Also, this would make `MakeNullScalar(union_type)` returning some slightly weird result with a `true` validity...

Reply:
Sorry for my incorrect `suggestion`. Indeed: if we wanted to maintain consistent "shape" there can be no such thing as a union scalar for which `is_valid == false` since that would be analogous to a union array with a top level null slot. Having `MakeNullSalar(union_type)->is_valid == true` is counterintuitive, but it is consistent with the behavior of union arrays: `MakeArrayOfNull(union_type, 1).ValueOrDie()->IsValid(0) == true` so I still think it's the correct decision here.

I do think the caveat is counterintuitive enough to explicitly note in the doccomments for MakeArrayOfNull and MakeNullScalar.

Reply:
This might be worth discussing on the ML, what do you think?

Reply:
SGTM, would you like me to write or will you?

Reply:
You can write if you have time, otherwise I'll do it tomorrow.